go/store/nbs/cmp_chunk_table_writer.go

```diff
@@ -0,0 +1,321 @@
+// data than the fixed buffer can hold will result in an error
+type FixedBufferByteSink struct {
+	buff []byte
+	pos  uint64
+}
+
+// NewFixedBufferTableSink creates a FixedBufferTableSink which will use the supplied buffer
+func NewFixedBufferTableSink(buff []byte) *FixedBufferByteSink {
+	if len(buff) == 0 {
+		panic("must provide a buffer")
+	}
+
+	return &FixedBufferByteSink{buff: buff}
+}
+
+// Write writes a byte array to the sink.
+func (sink *FixedBufferByteSink) Write(src []byte) (int, error) {
+	dest := sink.buff[sink.pos:]
+	destLen := len(dest)
+	srcLen := len(src)
+
+	if destLen < srcLen {
+		return 0, ErrBufferFull
+	}
+
+	n := copy(dest, src)
+
+	if n != srcLen {
+		return 0, ErrBufferFull
+	}
+
+	sink.pos += uint64(n)
+	return srcLen, nil
+}
+
+// Flush writes all the data that was written to the ByteSink to the supplied writer
+func (sink *FixedBufferByteSink) Flush(wr io.Writer) error {
+	return iohelp.WriteAll(wr, sink.buff[:sink.pos])
+}
+
+// BlockBufferByteSink allocates blocks of data which of a given block size to store the bytes written to the sink. New
+// blocks are allocated as needed in order to handle all the data of the Write calls.
+type BlockBufferByteSink struct {
+	blockSize int
+	pos       uint64
+	blocks    [][]byte
+}
+
+// NewBlockBufferTableSink creates a BlockBufferByteSink with the provided block size.
+func NewBlockBufferTableSink(blockSize int) *BlockBufferByteSink {
+	block := make([]byte, 0, blockSize)
+	return &BlockBufferByteSink{blockSize, 0, [][]byte{block}}
+}
+
+// Write writes a byte array to the sink.
+func (sink *BlockBufferByteSink) Write(src []byte) (int, error) {
+	srcLen := len(src)
+	currBlockIdx := len(sink.blocks) - 1
+	currBlock := sink.blocks[currBlockIdx]
+	remaining := cap(currBlock) - len(currBlock)
+
+	if remaining >= srcLen {
+		currBlock = append(currBlock, src...)
+		sink.blocks[currBlockIdx] = currBlock
+	} else {
+		if remaining > 0 {
+			currBlock = append(currBlock, src[:remaining]...)
+			sink.blocks[currBlockIdx] = currBlock
+		}
+
+		newBlock := make([]byte, 0, sink.blockSize)
+		newBlock = append(newBlock, src[remaining:]...)
+		sink.blocks = append(sink.blocks, newBlock)
+	}
+
+	sink.pos += uint64(srcLen)
+	return srcLen, nil
+}
+
+// Flush writes all the data that was written to the ByteSink to the supplied writer
+func (sink *BlockBufferByteSink) Flush(wr io.Writer) (err error) {
+	return iohelp.WriteAll(wr, sink.blocks...)
+}
+
+const defaultTableSinkBlockSize = 2 * 1024 * 1024
+
+// ErrNotFinished is an error returned by a CmpChunkTableWriter when a call to Flush* is called before Finish is called
+var ErrNotFinished = errors.New("Not finished")
+
+// ErrAlreadyFinished is an error returned if Finish is called more than once on a CmpChunkTableWriter
+var ErrAlreadyFinished = errors.New("Already Finished")
+
+// CmpChunkTableWriter writes CompressedChunks to a table file
+type CmpChunkTableWriter struct {
```

Comment:
Much of the IO code below here would benefit from compressing the error handling to inlined if statements where possible.

Reply:
I don't like that style